In [1]:
 # Dependencies and Setup
import pandas as pd
import numpy as np

In [2]:
# File to Load (Remember to Change These)
file = "pymoli.csv"

purchase_data = pd.read_csv(file)
# purchase_data.head()

In [3]:
# number of players
unique_players = purchase_data["SN"].unique()
total_players = len(unique_players)
total_players
player_count_df = pd.DataFrame(
    {"Total Players": [total_players]
    }
)


# Player Count

In [4]:
player_count_df

Total Players
0            576

In [5]:
unique_items = purchase_data["Item ID"].unique()
total_items = len(unique_items)

avg_purchase_price = purchase_data["Price"].mean()

purchases = purchase_data["Purchase ID"].count()

total_revenue = purchase_data["Price"].sum()

purchase_analysis_df = pd.DataFrame(
    {"Number of Unique Items": [total_items],
     "Average Price": [avg_purchase_price],
     "Number of Purchases": [purchases],
     "Total Revenue": [total_revenue]   
    })
purchase_analysis_df['Average Price']= purchase_analysis_df['Average Price'].map("${:.2f}".format)
purchase_analysis_df['Total Revenue']= purchase_analysis_df['Total Revenue'].map("${:,.2f}".format)


# Purchasing Analysis (Total)

In [6]:
purchase_analysis_df

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     183         $3.05                  780     $2,379.77

In [7]:
gen_df = purchase_data.drop_duplicates(subset=['SN'], keep = 'first')
gen_df.groupby(['Gender'])
num = len(purchase_data['SN'].unique())
num = pd.to_numeric(num)


In [8]:
males = gen_df.loc[gen_df['Gender']== 'Male',:]
females = gen_df.loc[gen_df['Gender']=='Female',:]
other = gen_df.loc[gen_df['Gender']== 'Other / Non-Disclosed',:]
male_count = len(males['Gender'])
female_count = len(females['Gender'])
other_count = len(other['Gender'])
male_per = (male_count/num)*100

female_per = (female_count/num)*100
other_per = (other_count/num)*100


In [9]:
gender_dem_df = pd.DataFrame(
    {"Total Count": [male_count, female_count, other_count],
     "Percentage of Players": [male_per, female_per, other_per],  
    })

gender_dem_df['']=['Male', 'Female', 'Other / Non-Disclosed']
gender_dem_df = gender_dem_df[['', 'Total Count', 'Percentage of Players']]
gender_dem_df['Percentage of Players']=gender_dem_df['Percentage of Players'].map("{:.2f}".format)

# Gender Demographics

In [10]:
gender_dem_df

Total Count Percentage of Players
0                   Male          484                 84.03
1                 Female           81                 14.06
2  Other / Non-Disclosed           11                  1.91

In [11]:
tot_male= purchase_data.loc[purchase_data['Gender']=='Male',:]
group_tot_male = tot_male.groupby(['SN']).sum().mean()
tot_female= purchase_data.loc[purchase_data['Gender']=='Female',:]
group_tot_female = tot_female.groupby(['SN']).sum().mean()
tot_other= purchase_data.loc[purchase_data['Gender']== 'Other / Non-Disclosed', :]
group_tot_other = tot_other.groupby(['SN']).sum().mean()

In [12]:
tot_male_count = len(tot_male['Purchase ID'])
tot_female_count = len(tot_female['Purchase ID'])
tot_other_count = len(tot_other['Purchase ID'])
male_avg_purch = tot_male['Price'].mean()
female_avg_purch = tot_female['Price'].mean()
other_avg_purch = tot_other['Price'].mean()
male_purch_value = tot_male['Price'].sum()
female_purch_value = tot_female['Price'].sum()
other_purch_value = tot_other['Price'].sum()


In [13]:

gen_purch_df = pd.DataFrame(
    {'Gender': ['Female', 'Male', 'Other / Non-Disclosed'],
     'Purchase Count': [tot_female_count, tot_male_count, tot_other_count],
     'Average Purchase Price': [female_avg_purch, male_avg_purch, other_avg_purch],
     'Total Purchase Value': [female_purch_value, male_purch_value, other_purch_value],
     'Avg Total Purchase per Person':[group_tot_female['Price'].mean(),group_tot_male['Price'].mean(),group_tot_other['Price'].mean()]
        
    })
gen_purch_df['Average Purchase Price']=gen_purch_df['Average Purchase Price'].map("${:.2f}".format)
gen_purch_df['Total Purchase Value']=gen_purch_df['Total Purchase Value'].map("${:,.2f}".format)
gen_purch_df['Avg Total Purchase per Person']=gen_purch_df['Avg Total Purchase per Person'].map("${:.2f}".format)

# Purchasing Analysis (Gender)

In [14]:
gen_purch_df

Gender  Purchase Count Average Purchase Price  \
0                 Female             113                  $3.20   
1                   Male             652                  $3.02   
2  Other / Non-Disclosed              15                  $3.35   

  Total Purchase Value Avg Total Purchase per Person  
0              $361.94                         $4.47  
1            $1,967.64                         $4.07  
2               $50.19                         $4.56

In [15]:
bins = [0, 9, 14, 19, 24, 29, 34, 39,150]
group_labels = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]
gen_df['Age Group'] = pd.cut(gen_df['Age'], bins, labels=group_labels)
purchase_data['Age Group'] = pd.cut(purchase_data['Age'], bins, labels=group_labels)
ten_young = gen_df.loc[gen_df['Age Group']=='<10']
ten_young = len(ten_young['Age'])

ten_fourteen= gen_df.loc[gen_df['Age Group']=='10-14']
ten_fourteen= len(ten_fourteen['Age'])

fif_nineteen = gen_df.loc[gen_df['Age Group']=='15-19']
fif_nineteen = len(fif_nineteen['Age'])

twen_twfour =  gen_df.loc[gen_df['Age Group']=='20-24']
twen_twfour = len(twen_twfour['Age'])

twfive_twnine =  gen_df.loc[gen_df['Age Group']=='25-29']
twfive_twnine = len(twfive_twnine['Age'])

thirty_thfour =  gen_df.loc[gen_df['Age Group']=='30-34']
thirty_thfour = len(thirty_thfour['Age'])

thfive_thnine =  gen_df.loc[gen_df['Age Group']=='35-39']
thfive_thnine = len(thfive_thnine['Age'])

forty_over =  gen_df.loc[gen_df['Age Group']=='40+']
forty_over = len(forty_over['Age'])

num_players= [ten_young, ten_fourteen, fif_nineteen, twen_twfour, twfive_twnine,\
             thirty_thfour, thfive_thnine, forty_over]
ten_young_per = (ten_young/num)*100
ten_fourteen_per = (ten_fourteen/num)*100
fif_nineteen_per = (fif_nineteen/num)*100
twen_twfour_per = (twen_twfour/num)*100
twfive_twnine_per = (twfive_twnine/num)*100
thirty_thfour_per = (thirty_thfour/num)*100
thfive_thnine_per = (thfive_thnine/num)*100
forty_over_per = (forty_over/num)*100
percentage_players= [ten_young_per, ten_fourteen_per, fif_nineteen_per, \
                    twen_twfour_per, twfive_twnine_per, thirty_thfour_per, \
                    thfive_thnine_per, forty_over_per]

C:\ProgramData\Anaconda3\envs\PythonData\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [16]:
age_dem_df = pd.DataFrame(
    {"Total Count": num_players,
     "Percentage of Players": percentage_players
    
})

age_dem_df['Percentage of Players']=age_dem_df['Percentage of Players'].map("{:.2f}".format)
age_dem_df= age_dem_df.set_index([group_labels], drop=True)

# Age Demographics

In [17]:
age_dem_df

Total Count Percentage of Players
<10             17                  2.95
10-14           22                  3.82
15-19          107                 18.58
20-24          258                 44.79
25-29           77                 13.37
30-34           52                  9.03
35-39           31                  5.38
40+             12                  2.08

In [18]:
young_df = purchase_data.loc[purchase_data['Age Group'] == '<10',:]
ten_four_df = purchase_data.loc[purchase_data['Age Group'] == '10-14',:]
fif_nin_df = purchase_data.loc[purchase_data['Age Group'] == '15-19',:]
twen_twfour_df = purchase_data.loc[purchase_data['Age Group'] == '20-24',:]
twfiv_twnin_df = purchase_data.loc[purchase_data['Age Group'] == '25-29',:]
thir_thfour_df = purchase_data.loc[purchase_data['Age Group'] == '30-34',:]
thfiv_thnin_df = purchase_data.loc[purchase_data['Age Group'] == '35-39',:]
forty_over_df = purchase_data.loc[purchase_data['Age Group'] == '40+',:]


In [19]:
grouped_young= young_df.groupby(['SN']).sum().mean()
grouped_ten= ten_four_df.groupby(['SN']).sum().mean()
grouped_fif= fif_nin_df.groupby(['SN']).sum().mean()
grouped_twen=twen_twfour_df.groupby(['SN']).sum().mean()
grouped_twfiv=twfiv_twnin_df.groupby(['SN']).sum().mean()
grouped_thir=thir_thfour_df.groupby(['SN']).sum().mean()
grouped_thfiv= thfiv_thnin_df.groupby(['SN']).sum().mean()
grouped_fort= forty_over_df.groupby(['SN']).sum().mean()

In [20]:
age_purch_df = pd.DataFrame(
    {"Purchase Count": [len(ten_four_df['Purchase ID']), len(fif_nin_df['Purchase ID']), \
                        len(twen_twfour_df['Purchase ID']), len(twfiv_twnin_df['Purchase ID']), \
                        len(thir_thfour_df['Purchase ID']), len(thfiv_thnin_df['Purchase ID']), \
                        len(forty_over_df['Purchase ID']), len(young_df['Purchase ID'])],
     
     "Average Purchase Price": [ten_four_df['Price'].mean(), fif_nin_df['Price'].mean(), \
                               twen_twfour_df['Price'].mean(), twfiv_twnin_df['Price'].mean(), \
                                thir_thfour_df['Price'].mean(), thfiv_thnin_df['Price'].mean(), \
                                forty_over_df['Price'].mean(), young_df['Price'].mean()],
     
     "Total Purchase Value": [ten_four_df['Price'].sum(), fif_nin_df['Price'].sum(), \
                             twen_twfour_df['Price'].sum(), twfiv_twnin_df['Price'].sum(), \
                             thir_thfour_df['Price'].sum(), thfiv_thnin_df['Price'].sum(), \
                             forty_over_df['Price'].sum(), young_df['Price'].sum()],
     "Avg Total Purchase per Person": [grouped_ten['Price'].mean(), grouped_fif['Price'].mean(), \
                                      grouped_twen['Price'].mean(), grouped_twfiv['Price'].mean(), \
                                      grouped_thir['Price'].mean(), grouped_thfiv['Price'].mean(), \
                                      grouped_fort['Price'].mean(), grouped_young['Price'].mean()]
})
age_purch_df = age_purch_df.set_index([group_labels], drop=True)
age_purch_df['Average Purchase Price']=age_purch_df['Average Purchase Price'].map("${:.2f}".format)
age_purch_df['Total Purchase Value']=age_purch_df['Total Purchase Value'].map("${:,.2f}".format)
age_purch_df['Avg Total Purchase per Person']=age_purch_df['Avg Total Purchase per Person'].map("${:.2f}".format)

# Purchasing Analysis (Age)

In [21]:
age_purch_df

Purchase Count Average Purchase Price Total Purchase Value  \
<10                28                  $2.96               $82.78   
10-14             136                  $3.04              $412.89   
15-19             365                  $3.05            $1,114.06   
20-24             101                  $2.90              $293.00   
25-29              73                  $2.93              $214.00   
30-34              41                  $3.60              $147.67   
35-39              13                  $2.94               $38.24   
40+                23                  $3.35               $77.13   

      Avg Total Purchase per Person  
<10                           $3.76  
10-14                         $3.86  
15-19                         $4.32  
20-24                         $3.81  
25-29                         $4.12  
30-34                         $4.76  
35-39                         $3.19  
40+                           $4.54

In [22]:
grouped_purchase_data = purchase_data.groupby('SN')
mean_grouped= grouped_purchase_data.mean()
sum_grouped = grouped_purchase_data.sum()
sum_grouped = sum_grouped.sort_values('Price', ascending=False)
sum_grouped = sum_grouped.head()
sum_grouped

Purchase ID  Age  Item ID  Price
SN                                           
Lisosia93           1630  125      451  18.96
Idastidru52         1999   96      527  15.45
Chamjask73          1306   66      339  13.83
Iral74              2285   84      518  13.62
Iskadarya95          713   60      321  13.10

In [23]:
# purchase_data = purchase_data.sort_values('Price', ascending=False)

Lis = purchase_data.loc[purchase_data['SN'] == 'Lisosia93',:]
Ida = purchase_data.loc[purchase_data['SN'] == 'Idastidru52',:]
Cha = purchase_data.loc[purchase_data['SN'] == 'Chamjask73',:]
Ira = purchase_data.loc[purchase_data['SN'] == 'Iral74',:]
Isk = purchase_data.loc[purchase_data['SN'] == 'Iskadarya95',:]

Lis_purch = len(Lis['SN'])
Ida_purch = len(Ida['SN'])
Cha_purch = len(Cha['SN'])
Ira_purch = len(Ira['SN'])
Isk_purch = len(Isk['SN'])
purch_list = [Lis_purch, Ida_purch, Cha_purch, Ira_purch, Isk_purch]

Lis_avg_purch = Lis['Price'].mean()
Ida_avg_purch = Ida['Price'].mean()
Cha_avg_purch = Cha['Price'].mean()
Ira_avg_purch = Ira['Price'].mean()
Isk_avg_purch = Isk['Price'].mean()
avg_purch_list = [Lis_avg_purch, Ida_avg_purch, Cha_avg_purch, Ira_avg_purch, Isk_avg_purch]

Lis_tot_purch = Lis['Price'].sum()
Ida_tot_purch = Ida['Price'].sum()
Cha_tot_purch = Cha['Price'].sum()
Ira_tot_purch = Ira['Price'].sum()
Isk_tot_purch = Isk['Price'].sum()
tot_purch_list = [Lis_tot_purch, Ida_tot_purch, Cha_tot_purch, Ira_tot_purch, Isk_tot_purch]

In [24]:
top_spender_df = pd.DataFrame(
    {"Purchase Count": purch_list, 
     "Average Purchase Price": avg_purch_list,
     "Total Purchase Value": tot_purch_list
        
    })
top_spenders = ["Lisosia93", "Idastidru52", "Chamjask73", "Iral74", "Iskadarya95"]
top_spender_df = top_spender_df.set_index([top_spenders], drop=True)
top_spender_df['Average Purchase Price']= top_spender_df['Average Purchase Price'].map("${:.2f}".format)
top_spender_df['Total Purchase Value']= top_spender_df['Total Purchase Value'].map("${:.2f}".format)

# Top Spenders

In [25]:
top_spender_df

Purchase Count Average Purchase Price Total Purchase Value
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

In [26]:
item_df = purchase_data[['Item ID', 'Item Name', 'Price']]

grouped_item_df = item_df.groupby(['Item ID', 'Item Name'])

pop_item_count = grouped_item_df.count().sort_values('Price', ascending = False).head()
print(pop_item_count)
pop_item_sum = grouped_item_df.sum()
pop_df = pd.merge(pop_item_count, pop_item_sum, on='Item Name', how= 'inner')
pop_price = item_df.drop_duplicates(subset = 'Item Name', keep= 'first')
oath = purchase_data.loc[purchase_data['Item ID']==178, 'Price'].max()
fier = purchase_data.loc[purchase_data['Item ID']==145, 'Price'].max()
extr = purchase_data.loc[purchase_data['Item ID']==108,'Price'].max()
nirv = purchase_data.loc[purchase_data['Item ID']==82, 'Price'].max()
purs = purchase_data.loc[purchase_data['Item ID']==19, 'Price'].max()

price_list= [oath, fier, extr, nirv, purs]

                                                      Price
Item ID Item Name                                          
178     Oathbreaker, Last Hope of the Breaking Storm     12
145     Fiery Glass Crusader                              9
108     Extraction, Quickblade Of Trembling Hands         9
82      Nirvana                                           9
19      Pursuit, Cudgel of Necromancy                     8


In [27]:
pop_df['Item Price'] = price_list
pop_df =pop_df.rename(columns = {'Price_x': 'Purchase Count', 'Price_y': 'Total Purchase Value'})
pop_df['Total Purchase Value']= pop_df['Total Purchase Value'].map("${:.2f}".format)
pop_df['Item Price']= pop_df['Item Price'].map("${:.2f}".format)
columntit= ['Purchase Count', 'Item Price', 'Total Purchase Value']
pop_df= pop_df.reindex(columns=columntit)

# Most Popular Items

In [28]:
pop_df

Purchase Count Item Price  \
Item Name                                                                 
Oathbreaker, Last Hope of the Breaking Storm              12      $4.23   
Fiery Glass Crusader                                       9      $4.58   
Extraction, Quickblade Of Trembling Hands                  9      $3.53   
Nirvana                                                    9      $4.90   
Pursuit, Cudgel of Necromancy                              8      $1.02   

                                             Total Purchase Value  
Item Name                                                          
Oathbreaker, Last Hope of the Breaking Storm               $50.76  
Fiery Glass Crusader                                       $41.22  
Extraction, Quickblade Of Trembling Hands                  $31.77  
Nirvana                                                    $44.10  
Pursuit, Cudgel of Necromancy                               $8.16

In [29]:
prof_df= pop_item_sum.sort_values('Price', ascending=False)
prof_df=prof_df.head()



In [30]:
oath = purchase_data.loc[purchase_data['Item ID']==178, 'Price'].max()
nirv = purchase_data.loc[purchase_data['Item ID']==82, 'Price'].max()
fier = purchase_data.loc[purchase_data['Item ID']==145, 'Price'].max()
fina = purchase_data.loc[purchase_data['Item ID']== 92, 'Price'].max()
sing = purchase_data.loc[purchase_data['Item ID']==103, 'Price'].max()
prof_price_list = [oath, nirv, fier, fina, sing]

In [31]:
oath = purchase_data.loc[purchase_data['Item ID']==178, 'Price'].count()
nirv = purchase_data.loc[purchase_data['Item ID']==82, 'Price'].count()
fier = purchase_data.loc[purchase_data['Item ID']==145, 'Price'].count()
fina = purchase_data.loc[purchase_data['Item ID']==92, 'SN'].count()
sing = purchase_data.loc[purchase_data['Item ID']==103, 'Price'].count()
prof_count_list = [oath, nirv, fier, fina, sing]

In [32]:
prof_df['Purchase Count']= prof_count_list
prof_df['Item Price']= prof_price_list

In [33]:
prof_df = prof_df.rename(columns={'Price': 'Total Purchase Value'})
prof_df['Total Purchase Value']=prof_df['Total Purchase Value'].map("${:.2f}".format)
prof_df['Item Price']=prof_df['Item Price'].map("${:.2f}".format)
columnstitles= ['Purchase Count', 'Item Price', 'Total Purchase Value']
prof_df= prof_df.reindex(columns=columnstitles)

# Most Profitable Items

In [34]:
prof_df

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
92      Final Critic                                               8   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
92      Final Critic                                      $4.88   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
92      Final Critic                                               $39.04  
103     Singed Scalpel                                             $34.80

# 3 Observable Trends

1. The top spenders make at least 3 purchases.

2. Players ages 15-19 had the highest total purchase value of 1,114.06 dollars and out spent the next closest age group (10-14) by 700 dollars.  

3. The most popular and most profitable item was the Oathbreaker, Last Hope of the Breaking Storm. It was purchased 12 times and had a total purchase value of 50.76 dollars.